enhancements/authentication/direct-external-oidc-provider.md

```diff
@@ -212,6 +212,90 @@ These changes have already been implemented, and the initial PR for them can be
+- The post-start hook that checks for oauth-apiserver connectivity will be skipped if the `Authentication` resource `.Spec.Type` is set to `OIDC`
+- `RoleBinding`s will be rejected if there exists a `RoleBindingRestriction` that specifies user and/or group restrictions in the namespace the `RoleBinding` is being created
+- It is considered a failure if we are unable to determine the authentication type for the cluster, leading to rejection of the `RoleBinding`
+
+##### Changes to openshift-apiserver
+
+To help keep users informed of the expected behavior of the `authorization.openshift.io/RestrictSubjectBindings` admission plugin when using the OIDC cluster authentication mode, it is proposed that a new admission plugin is added to the openshift-apiserver to reject creation of `RoleBindingRestriction` resources containing user/group restrictions.
+
+**Alternatives**
+
+- Do not reject admission, but issue a warning of the impacts creating a `RoleBindingRestriction` may have when using OIDC as the cluster authentication method.
+- Use a `ValidatingAdmissionPolicy` + `ValidatingAdmissionPolicyBinding` instead of an admission plugin.
+- Don't do anything special, let changes in the `authorization.openshift.io/RestrictSubjectBindings` admission plugin and documentation be the source of this information.
+
+##### Changes to the `Authentication` API
+
+The OIDC authentication mode on the cluster will not be enabled if there exists `RoleBindingRestriction` resources that specify user/group restrictions.
+
+To communicate the reason for the enablement of the OIDC functionality being blocked, the `Authentication` API will be extended with a new status field to communicate the condition of the OIDC feature.
+
+```go
+type AuthenticationStatus struct {
+    ...
+
+    // OIDCConditions are conditions related to the
+    // enablement of the OIDC authentication feature.
+    //
+    // The Enabled condition represents whether or not the OIDC
+    // feature has been successfully enabled.
+    //
+    // When Enabled is False and the reason is RoleBindingRestrictionViolations,
+    // the rollout of the OIDC feature could not progress due to the existence
+    // of RoleBindingRestrictions that specify user and/or group subject restrictions.
+    // Enabling the OIDC functionality removes the User and Group APIs from the cluster
+    // and as such, RoleBindingRestrictions should not specify user or group restrictions
+    // when using an OIDC provider for authentication.
+    //
+    // When Enabled is True and the reason is Succeeded, the rollout of the OIDC feature
+    // has been successful and your cluster is officially using the OIDC provider
+    // for authentication.
+    //
+    // +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+    // +openshift:enable:FeatureGate=ExternalOIDC
+    OIDCConditions []metav1.Condition `json:"oidcConditions"`
```

Comment:
Here's an idea of how we could structure this to make it usable for more generic tracking of the OIDC config/rollout progress:

```yaml
status:
  oidcConfig:
    conditions:
    - message: ""
      reason: ""
      status: "True" # use this to indicate full KAS rollout on OIDC and start cleanup
      type: "Available"
    - message: ""
      reason: ""
      status: "False" # use this to indicate rollout in progress (new & updated config)
      type: "Progressing"
    - message: ""
      reason: ""
      status: "True" # use this to indicate CAO/KAS-o configuration issues
      type: "Degraded"
    # just an example structure about RBRs -- a separate condition type would allow to split meaning from the standard Available/Progressing/Degraded
    - message: "existing RoleBindingRestrictions on users/group not supported in external OIDC: ns1/rbr1, ns2/rbr2"
      reason: "UnsupportedRoleBindingRestrictions"
      status: "True"
      type: "UnsupportedResourceDetected"
```

It would be useful to separate this into `status.oidcConfig.conditions` because we might want to track other fields under `status.oidcConfig` (similar to `status.oidcClients`).

Finally, if we decide to block enablement if there are RBRs on Users/Groups, instead of a dedicated condition type we would use `Degraded`.